### PR TITLE
Add Echo State Network (ESN) layer

### PR DESCRIPTION
**Pull request by pedrolarben**
Part of https://github.com/tensorflow/addons/pull/1809.

After adding the ESN recurrent cell https://github.com/tensorflow/addons/pull/1811, here it is the ESN layer.

### CODE REVIEW

File: tensorflow_addons/layers/esn.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+class ESN(tf.keras.layers.RNN):
+    """Echo State Network layer.
+
+    This implements the recurrent layer using the ESNCell.
+
+    This is based on the paper
+        H. Jaeger
+        "The "echo state" approach to analysing and training recurrent neural networks".
+        GMD Report148, German National Research Center for Information Technology, 2001.
+        https://www.researchgate.net/publication/215385037
```
Comment (WindQAQ) [inline tensorflow_addons/layers/esn.py:39]: Could we use the hyperlink here? Like [xxx](link).

File: tensorflow_addons/layers/esn.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+
```
Comment (WindQAQ) [inline tensorflow_addons/layers/esn.py:41]: Please remove the empty line here.

File: tensorflow_addons/layers/esn.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            spectral_radius=spectral_radius,
+            use_norm2=use_norm2,
+            use_bias=use_bias,
+            activation=activation,
+            kernel_initializer=kernel_initializer,
+            recurrent_initializer=recurrent_initializer,
+            bias_initializer=bias_initializer,
+            dtype=kwargs.get("dtype"),
+        )
+        super(ESN, self).__init__(
```
Comment (WindQAQ) [inline tensorflow_addons/layers/esn.py:127]: We drop the support to python2 so using `super().__init__` is enough.

File: tensorflow_addons/layers/esn.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+            return_sequences=return_sequences,
+            go_backwards=go_backwards,
+            unroll=unroll,
+            **kwargs,
+        )
+
+    def call(
+        self, inputs, mask=None, training=None, initial_state=None, constants=None
+    ):
+        return super(ESN, self).call(
```
Comment (WindQAQ) [inline tensorflow_addons/layers/esn.py:138]: ditto

File: tensorflow_addons/layers/esn.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+            "activation": tf.keras.activations.serialize(self.activation),
+            "kernel_initializer": tf.keras.initializers.serialize(
+                self.kernel_initializer
+            ),
+            "recurrent_initializer": tf.keras.initializers.serialize(
+                self.recurrent_initializer
+            ),
+            "bias_initializer": tf.keras.initializers.serialize(self.bias_initializer),
+        }
+        base_config = super(ESN, self).get_config()
```
Comment (WindQAQ) [inline tensorflow_addons/layers/esn.py:203]: ditto

File: tensorflow_addons/layers/tests/esn_test.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import pytest
+import numpy as np
+import tensorflow as tf
+from tensorflow_addons.layers.esn import ESN
+from tensorflow_addons.utils import test_utils
+
+
+@pytest.mark.usefixtures("maybe_run_functions_eagerly")
+@pytest.mark.parametrize("dtype", [np.float16, np.float32, np.float64])
+def test_random(dtype):
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/esn_test.py:26]: NIT: this is not a test case about randomized input :smiley: Maybe we could change its name.

File: tensorflow_addons/layers/esn.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        super(ESN, self).__init__(
+            cell,
+            return_sequences=return_sequences,
+            go_backwards=go_backwards,
+            unroll=unroll,
+            **kwargs,
+        )
+
+    def call(
+        self, inputs, mask=None, training=None, initial_state=None, constants=None
```
Comment (WindQAQ) [inline tensorflow_addons/layers/esn.py:136]: If `ESN` is not supporting `constants`, I think we could drop it like what `LSTM` does.

https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/layers/recurrent_v2.py#L1104

File: tensorflow_addons/layers/esn.py (hunk 1 above)
Comment (pedrolarben) [reply tensorflow_addons/layers/esn.py:39]: @WindQAQ Thanks for your review :)

Do you mean something like this? 
_Option A:_
 ```
This is based on the paper
H. Jaeger
["The "echo state" approach to analysing and training recurrent neural networks".
GMD Report148, German National Research Center for Information Technology, 2001.]
(https://www.researchgate.net/publication/215385037)
```
_Option B:_
 ```
This is based on the [paper](https://www.researchgate.net/publication/215385037)
H. Jaeger
"The "echo state" approach to analysing and training recurrent neural networks".
GMD Report148, German National Research Center for Information Technology, 2001.
```
_Option C:_
 ```
This is based on the paper
H. Jaeger
["The "echo state" approach to analysing and training recurrent neural networks"]
(https://www.researchgate.net/publication/215385037).
GMD Report148, German National Research Center for Information Technology, 2001.
```

I think I prefer option C, please let me know your opinion

File: tensorflow_addons/layers/esn.py (hunk 2 above)
Comment (pedrolarben) [reply tensorflow_addons/layers/esn.py]: Done!

File: tensorflow_addons/layers/esn.py (hunk 3 above)
Comment (pedrolarben) [reply tensorflow_addons/layers/esn.py]: Done!

File: tensorflow_addons/layers/esn.py (hunk 4 above)
Comment (pedrolarben) [reply tensorflow_addons/layers/esn.py]: Done!

File: tensorflow_addons/layers/esn.py (hunk 5 above)
Comment (pedrolarben) [reply tensorflow_addons/layers/esn.py]: Done!

File: tensorflow_addons/layers/tests/esn_test.py (hunk 6 above)
Comment (pedrolarben) [reply tensorflow_addons/layers/tests/esn_test.py]: Done 😅

File: tensorflow_addons/layers/esn.py (hunk 7 above)
Comment (pedrolarben) [reply tensorflow_addons/layers/esn.py]: I agree, Done!

File: tensorflow_addons/layers/esn.py (hunk 1 above)
Comment (WindQAQ) [reply tensorflow_addons/layers/esn.py:39]: @pedrolarben I prefer C too :-)

Comment (pedrolarben) [reply tensorflow_addons/layers/esn.py]: Done! :+1:

**Pull request merged**